README.md
Bloom Server
============

Bloom Filter Server - by btnguyen2k.

Project home: [https://github.com/btnguyen2k/bloom-server](https://github.com/btnguyen2k/bloom-server).

## Overview ##

Bloom Server provides REST and Thrift APIs to access centralized [bloom filter](http://en.wikipedia.org/wiki/Bloom_filter) data.


## Release-notes ##

Latest release: `0.1.1`.

See [RELEASE-NOTES.md](RELEASE-NOTES.md).


## APIs ##

* `put(item[, bloomName])`: put an item to a bloom filter.
* `mightContain(item[, bloomname])`: test if an item has been put to a bloom filter.
* `initBloom(bloomName, force, expectedNumItems, expectedFpp)`: create and initialize a bloom filter.

### REST APIs ###

* `GET /put/<item>[/bloomName]`
  - Put an item to a bloom filter.
  - `item (String)`: the item to be put to the bloom filter
  - `bloomName (String)`: (optional) name of the bloom filter to put the item to
  - Output: JSON `{"s":500,"v":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"v":false,"m":"Error message: the specified bloom filter does not exist"}`
  - Output: JSON `{"s":200,"v":true,"m":"Successful: bloom filter data is updated"}`
  - Output: JSON `{"s":200,"v":false,"m":"Successful: bloom filter has not changed, the item might have been put to the bloom filter before"}`

* `GET /mightContain/<item>[/bloomName]`
  - Test if an item has been put to a bloom filter.
  - `item (String)`: the item to be tested
  - `bloomName (String)`: (optional) name of the bloom filter to test
  - Output: JSON `{"s":500,"v":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":404,"v":false,"m":"Error message: the specified bloom filter does not exist"}`
  - Output: JSON `{"s":200,"v":false,"m":"Successful: the item has definitely not put to the bloom filter"}`
  - Output: JSON `{"s":200,"v":true,"m":"Successful: the item might has been put to the bloom filter"}`
  
* `POST /initBloom`
  - Creates and Initializes a new bloom filter.
  - Input: (JSON) `{"bloom_name":"(string) name of the bloom filter to create/init", "num_items":(int) number of expected items, "expected_fpp": (double, in range [0,1]) expected false-positive-probability, "force": (boolean) set to true to force override existing bloom filter, false to ignore if the bloom filter already existed, "secret": "(string) secret key to authorize the request"}`
  - Output: JSON `{"s":500,"v":false,"m":"Error message: exception occurred at server side"}`
  - Output: JSON `{"s":403,"v":false,"m":"Error message: unauthorized request (wrong secret code)"}`
  - Output: JSON `{"s":403,"v":false,"m":"Error message: unauthorized request (wrong secret code)"}`
  - Output: JSON `{"s":200,"v":true,"m":"Successful: new bloom filter has been created and initialized"}`
  - Output: JSON `{"s":200,"v":false,"m":"Successful: the bloom filter has already existed"}`

### Thrift APIs ###

See file [bloomserver.thrift](thrift/bloomserver.thrift).

### Thrift-over-HTTP APIs ###

Thrift APIs, but over HTTP(s)! URL: `http://server:host/thrift`

### Clients ###

- Java: [https://github.com/btnguyen2k/bloom-jclient](https://github.com/btnguyen2k/bloom-jclient).


## Installation ##

Note: Java 7+ is required!

### Install from binary ###

- Download binary package from [project release site](https://github.com/btnguyen2k/bloom-server/releases).
- Unzip the binary package and copy it to your favourite location, e.g. `/usr/local/bloom-server`.


### Install from source ###

- Download application's source, either cloning github project or download the source package from [project release site](https://github.com/btnguyen2k/bloom-server/releases).
- Build [Play! Framework](https://www.playframework.com): `play dist`.
- The built binary package is available at `target/universal/bloom-server-<version>.zip`. You may copy it to your favourite location, e.g. `/usr/local/bloom-server`.


## Start/Stop Bloom-Server ##

### Linux ###

Start server with default options:
> `/usr/local/bloom-server/conf/server-production.sh start`

Start server with 1024M memory limit, REST & Thrift-over-HTTP APIs on port 18080, Thrift APIs on port 19090
> `/usr/local/bloom-server/conf/server-production.sh start -m 1024 -p 18080 -t 19090`
Set thrift port to 0 (`-t 0`) to disable Thrift APIs.

Default port for REST & Thrift-over-HTTP requests is `8080` and `9090` is the default port for Thrift requests (both port numbers are configurable).

Stop server:
> `/usr/local/bloom-server/conf/server-production.sh stop`


## Configurations ##

See file(s) `conf/application.conf` and `conf/spring/beans.xml`.


## License ##

See [LICENSE.txt](LICENSE.txt) for details. Copyright (c) 2015 btnguyen2k.

Third party libraries are distributed under their own license(s).


RELEASE-NOTES.md
Bloom Server Release Notes
==========================

Bloom Filter Server - by btnguyen2k.

2015-08-21: v0.1.1
------------------

- Expire cached bloom filters when out-of-memory.


2015-05-31: v0.1.0
------------------
First release:

- REST & Thrift APIs


app/controllers/Application.java
package controllers;

import globals.Registry;

import java.util.HashMap;
import java.util.Map;

import org.apache.commons.lang3.StringUtils;

import play.Logger;
import play.api.templates.Html;
import play.mvc.Http.RawBuffer;
import play.mvc.Http.RequestBody;
import play.mvc.Result;
import util.Constants;
import api.BloomApi;

import com.fasterxml.jackson.databind.JsonNode;
import com.github.ddth.commons.utils.DPathUtils;
import com.github.ddth.commons.utils.SerializationUtils;
import com.github.ddth.plommon.utils.PlayAppUtils;
import com.github.ddth.tsc.DataPoint;
import com.github.ddth.tsc.ICounter;
import com.github.ddth.tsc.ICounterFactory;

public class Application extends BaseController {

    private static Result doResponse(int status, boolean value, String message) {
        Map<String, Object> result = new HashMap<String, Object>();
        result.put(Constants.RESPONSE_FIELD_STATUS, status);
        result.put(Constants.RESPONSE_FIELD_VALUE, value);
        result.put(Constants.RESPONSE_FIELD_MESSAGE, message);
        response().setHeader(CONTENT_TYPE, "application/json");
        response().setHeader(CONTENT_ENCODING, "utf-8");
        Registry.updateCounters(status);
        return ok(SerializationUtils.toJsonString(result));
    }

    /*
     * Handles: GET:/put/:item/:bloomName
     */
    public static Result put(final String item, final String bloomName) {
        BloomApi bloomApi = Registry.getBloomApi();
        try {
            int result = bloomApi.put(bloomName, item != null ? item.getBytes(Constants.UTF8)
                    : Constants.EMPTY);
            switch (result) {
            case Constants.API_RESULT_BLOOM_NOTFOUND:
                return doResponse(404, false, "Bloom filter [" + bloomName + "] does not exist!");
            case Constants.API_RESULT_INVALID_PARAM:
                return doResponse(400, false, "Invalid item: " + item);
            case Constants.API_RESULT_FALSE:
                return doResponse(200, false, "Successful");
            case Constants.API_RESULT_TRUE:
                return doResponse(200, true, "Successful");
            }
            return doResponse(500, false, "Error");
        } catch (Exception e) {
            final String logMsg = "Exception [" + e.getClass() + "]: " + e.getMessage();
            Logger.error(logMsg, e);
            return doResponse(500, false, logMsg);
        }
    }

    /*
     * Handles: GET:/mightContain/:item/:bloomname
     */
    public static Result mightContain(final String item, final String bloomName) {
        BloomApi bloomApi = Registry.getBloomApi();
        try {
            int result = bloomApi.mightContain(bloomName,
                    item != null ? item.getBytes(Constants.UTF8) : Constants.EMPTY);
            switch (result) {
            case Constants.API_RESULT_BLOOM_NOTFOUND:
                return doResponse(404, false, "Bloom filter [" + bloomName + "] does not exist!");
            case Constants.API_RESULT_INVALID_PARAM:
                return doResponse(400, false, "Invalid item: " + item);
            case Constants.API_RESULT_FALSE:
                return doResponse(200, false, "Successful");
            case Constants.API_RESULT_TRUE:
                return doResponse(200, true, "Successful");
            }
            return doResponse(500, false, "Error");
        } catch (Exception e) {
            final String logMsg = "Exception [" + e.getClass() + "]: " + e.getMessage();
            Logger.error(logMsg, e);
            return doResponse(500, false, logMsg);
        }
    }

    /*
     * Handles: POST:/initBloom
     */
    @SuppressWarnings("unchecked")
    public static Result initBloom() {
        RequestBody requestBody = request().body();
        String requestContent = null;
        JsonNode jsonNode = requestBody.asJson();
        if (jsonNode != null) {
            requestContent = jsonNode.toString();
        } else {
            RawBuffer rawBuffer = requestBody.asRaw();
            if (rawBuffer != null) {
                requestContent = new String(rawBuffer.asBytes(), Constants.UTF8);
            } else {
                requestContent = requestBody.asText();
            }
        }

        Map<String, Object> params = SerializationUtils.fromJsonString(requestContent, Map.class);
        String bloomName = DPathUtils.getValue(params, "bloom_name", String.class);
        Long expectedNumItems = DPathUtils.getValue(params, "num_items", Long.class);
        if (expectedNumItems == null || expectedNumItems.longValue() < 1) {
            expectedNumItems = Constants.DEFAULT_EXPECTED_NUM_ITEMS;
        }
        Double expectedFpp = DPathUtils.getValue(params, "expected_fpp", Double.class);
        if (expectedFpp == null || expectedFpp.doubleValue() <= 0 || expectedFpp.doubleValue() >= 1) {
            expectedFpp = Constants.DEFAULT_EXPECTED_FPP;
        }
        Boolean force = DPathUtils.getValue(params, "force", Boolean.class);
        if (force == null) {
            force = Boolean.FALSE;
        }
        String secret = DPathUtils.getValue(params, "secret", String.class);

        // authorize the request
        String mySecret = PlayAppUtils.appConfigString("bloom.secret");
        if (StringUtils.isBlank(mySecret) || !StringUtils.equals(secret, mySecret)) {
            return doResponse(403, false, "Unauthorized request!");
        }

        BloomApi bloomApi = Registry.getBloomApi();
        try {
            boolean result = bloomApi.initBloomFilter(bloomName, force.booleanValue(),
                    expectedNumItems.longValue(), expectedFpp.doubleValue());
            Map<String, Object> resultData = new HashMap<String, Object>();
            resultData.put("num_items", expectedNumItems.longValue());
            resultData.put("fpp", expectedFpp.doubleValue());
            return doResponse(200, result, resultData.toString());
        } catch (Exception e) {
            final String logMsg = "Exception [" + e.getClass() + "]: " + e.getMessage();
            Logger.error(logMsg, e);
            return doResponse(500, false, logMsg);
        }
    }

    private static DataPoint[] buildCounterData(ICounter counter, long timestamp) {
        long last1Min = timestamp - 60 * 1000L;
        long last5Mins = timestamp - 5 * 60 * 1000L;
        long last15Mins = timestamp - 15 * 60 * 1000L;
        DataPoint[] result = new DataPoint[] {
                new DataPoint(DataPoint.Type.SUM, last1Min, 0, ICounter.STEPS_1_MIN * 1000),
                new DataPoint(DataPoint.Type.SUM, last5Mins, 0, ICounter.STEPS_5_MINS * 1000),
                new DataPoint(DataPoint.Type.SUM, last15Mins, 0, ICounter.STEPS_15_MINS * 1000) };
        if (counter == null) {
            return result;
        }

        DataPoint[] tempArr = counter.getSeries(last1Min, timestamp);
        for (DataPoint dp : tempArr) {
            result[0].add(dp);
        }
        tempArr = counter.getSeries(last5Mins, timestamp);
        for (DataPoint dp : tempArr) {
            result[1].add(dp);
        }
        tempArr = counter.getSeries(last15Mins, timestamp);
        for (DataPoint dp : tempArr) {
            result[2].add(dp);
        }
        return result;
    }

    /*
     * Handle: GET:/index
     */
    public static Result index() throws Exception {
        Map<String, DataPoint[]> statsLocal = new HashMap<String, DataPoint[]>();
        Map<String, DataPoint[]> statsGlobal = new HashMap<String, DataPoint[]>();
        Map<String, Long> countersLocal = new HashMap<String, Long>();
        Map<String, Long> countersGlobal = new HashMap<String, Long>();
        ICounterFactory localCounterFactory = Registry.getLocalCounterFactory();
        ICounterFactory globalCounterFactory = Registry.getGlobalCounterFactory();

        long timestamp = System.currentTimeMillis();
        String[] tscNames = new String[] { Registry.TSC_TOTAL, Registry.TSC_200, Registry.TSC_400,
                Registry.TSC_403, Registry.TSC_404, Registry.TSC_500 };
        for (String name : tscNames) {
            statsLocal.put(
                    name,
                    buildCounterData(
                            localCounterFactory != null ? localCounterFactory.getCounter(name)
                                    : null, timestamp));
            statsGlobal.put(
                    name,
                    buildCounterData(
                            globalCounterFactory != null ? globalCounterFactory.getCounter(name)
                                    : null, timestamp));
        }

        String[] counterNames = new String[] { Registry.COUNTER_TOTAL, Registry.COUNTER_200,
                Registry.COUNTER_400, Registry.COUNTER_403, Registry.COUNTER_404,
                Registry.COUNTER_500 };
        for (String name : counterNames) {
            ICounter counter = localCounterFactory != null ? localCounterFactory.getCounter(name)
                    : null;
            DataPoint dp = counter != null ? counter.get(0) : null;
            long value = dp != null ? dp.value() : 0;
            countersLocal.put(name, value);

            counter = globalCounterFactory != null ? globalCounterFactory.getCounter(name) : null;
            dp = counter != null ? counter.get(0) : null;
            value = dp != null ? dp.value() : 0;
            countersGlobal.put(name, value);
        }

        long[] concurrency = Registry.getConcurrency();

        Html html = render("index", concurrency, countersLocal, statsLocal, countersGlobal,
                statsGlobal);
        return ok(html);
    }

}
